### PR TITLE
build with mpi, petsc

### LINKED ISSUE #2: enable MPI
now that we've got a feedstock, we should enable building with MPI

### LINKED ISSUE #3: build with PETSc
once PETSc is building, we should link dolfin with PETSc

https://github.com/conda-forge/staged-recipes/pull/1488

### PR DESCRIPTION
**Pull request by minrk**
MPI runs aren't useful yet because no parallel graph partitioner is enabled. Neither PT-Scotch nor ParMetis is packaged on conda-forge yet, so will need to wait for that.

closes #2
closes #3

### CODE REVIEW

**Comment by garth-wells** [comment]
Using a packaged ParMETIS or PT-SCOTCH while likely be problematic. Much better is to let PETSc build a partitioner. Upstream maintenance of ParMETIS or PT-SCOTCH is not great, so PETSc team make sure to pick compatible versions and patch where required.

**Comment by minrk** [comment]
How do I find out which packages are well behaved and which should be bundled with petsc? I'd like to avoid bundling dependencies where I can, since each bundled dep produces a confict with the normal package. I'm about to open a PR on PETSc to expand the default set. My current plan is:

enable (not bundled):
- metis
- suitesparse

bundle:
- mumps
- parmetis

Anything I should change/add there?

**Comment by minrk** [comment]
If PETSc is just particular about versions, we can have regular dependencies and pin versions firmly. If it needs special configuration and/or patches, then bundling could be the way to go.

**Comment by garth-wells** [comment]
Unfortunately the root problem is poor upstream maintenance for the partitioners, and in some case the int size (this is a configure time options for some libraries). I would add

- Suitesparse/UMFPACK (this would probably from a package)
- MUMPS

ParMETIS screws up the licensing since it doesn't have an open-source license.

**Comment by mikaem** [comment]
I also bundle hypre. 
Do you think licensing is the reason why Parmetis is not on conda-forge, whereas metis is?

**Comment by garth-wells** [comment]
Likely. Metis changes to an open source license recently. ParMETIS page is down at the moment so I can't check latest licensing.

When PETSc builds MUMPS it might pull in ParMETIS. To keep it clean, we might want SuperLU_dist in place of MUMPS.

**Comment by garth-wells** [comment]
I'm happy to merge this to keep us moving towards an MPI-enabled build.

**Comment by minrk** [comment]
I've got some fiddling to do with scotch, I think I'll have something tomorrow after https://github.com/conda-forge/petsc-feedstock/pull/1 lands.

**Comment by garth-wells** [comment]
@minrk I think I introduced the merge conflicts. Could you fix manually in your branch?

**Comment by minrk** [comment]
I've fixed this locally. I'll push once some upstream builds finish propagating

**Comment by minrk** [comment]
This should be working now. CI may need to restart after the latest petsc builds finish.

**Comment by minrk** [comment]
CI passes, woo!

File: recipe/meta.yaml [hunk 1]
```diff
@@ -37,8 +44,13 @@ requirements:
     - cmake
     - eigen
     - gcc  # [linux]
+    - mpich
+    - petsc {{petsc_version}}
+    - petsc4py {{petsc4py_version}}
     - pkg-config
+    - ptscotch
     - zlib 1.2.*
+    - zlib
```
Comment (mikaem) [inline recipe/meta.yaml]: This line should be deleted because of the pinned version that is already there.

**Comment by mikaem** [comment]
Nice!

File: recipe/meta.yaml (hunk 1 above)
Comment (minrk) [inline recipe/meta.yaml]: Arg, failed conflict resolution. Fixed.

**Pull request merged**